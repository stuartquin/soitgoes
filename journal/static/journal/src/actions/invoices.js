'use strict';
import { push } from 'react-router-redux';
import { normalize } from 'normalizr';

import * as api from '../services/api';
import constants from '../constants';
import * as schema from '../actions/schema';

export const addItem = (invoiceId, name, price) => (dispatch) =>
  api.addInvoiceItem(invoiceId, name, price).then(item => {
    dispatch({
      type: constants.ADD_INVOICE_ITEM_SUCCESS,
      item
    });
  });

export const markAsIssued = (invoiceId, projectId, timeslips) => (dispatch) =>
  api.issueInvoice(invoiceId, projectId, timeslips).then(invoice => {
    dispatch({
      type: constants.SAVE_INVOICE_SUCCESS,
      invoice
    });
  });

export const markAsPaid = (invoiceId, projectId, totalPaid) => (dispatch) =>
  api.paidInvoice(invoiceId, projectId, totalPaid).then(invoice => {
    dispatch({
      type: constants.SAVE_INVOICE_SUCCESS,
      invoice
    });
  });

export const deleteInvoice = (invoiceId) => (dispatch) =>
  api.deleteInvoice(invoiceId).then(() => {
    dispatch(push(`/invoices`));
  });

export const createInvoice = (project) => (dispatch) => {
  dispatch({
    type: constants.CLEAR_INVOICE_TIMESLIPS
  });
  api.createInvoice(project).then(invoice => {
    dispatch({
      type: constants.CREATE_INVOICE_SUCCESS,
      invoice
    });
    dispatch(push(`/invoices/${invoice.id}`));
  });
};

export const fetchInvoice = (invoiceId) => (dispatch) => {
  return api.fetchInvoice(invoiceId).then(invoice =>
    dispatch({
      type: constants.GET_INVOICE_SUCCESS,
      invoice
    })
  );
};

export const fetchInvoices = () => (dispatch) => {
  dispatch({
    type: constants.CLEAR_INVOICE_TIMESLIPS
  });
  api.fetchInvoices().then(invoices =>
    dispatch({
      type: constants.GET_INVOICES_SUCCESS,
      invoices
    })
  );
};

export const fetchInvoiceTimeslips = (invoiceId) => (dispatch) =>
  api.fetchTimeslips(invoiceId).then(res => {
    dispatch({
      type: constants.GET_INVOICE_TIMESLIPS_SUCCESS,
      timeslips: res.results
    });
  });

export const fetchProjectTimeslips = (projectId) => (dispatch) =>
  api.fetchTimeslips('none', projectId).then(res => {
    dispatch({
      type: constants.GET_INVOICE_TIMESLIPS_SUCCESS,
      timeslips: res.resuslts
    });
  });
