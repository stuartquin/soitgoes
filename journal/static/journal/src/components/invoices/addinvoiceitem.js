'use strict';
import React from 'react';

import styles from './styles.css';

class AddInvoiceItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      costPerUnit: '',
      name: ''
    };
  }

  updateField(field, value) {
    let state = {};
    state[field] = value;
    this.setState(state);
  }

  render() {
    return (
      <tr>
        <td>
          <input
            placeholder='Additional Item'
            value={this.state.name}
            className={styles.invoiceItemText}
            onChange={(e) => this.updateField('name', e.target.value)}
            type='text' />
        </td>
        <td>
          <input
            placeholder='Price'
            value={this.state.costPerUnit}
            className={styles.invoiceItemNumber}
            onChange={(e) => this.updateField('costPerUnit', e.target.value)}
            type='number' />
        </td>
        <td>
          <input
            placeholder='Qty'
            value={this.state.qty}
            className={styles.invoiceItemNumber}
            onChange={(e) => this.updateField('qty', e.target.value)}
            type='number' />
        </td>
        <td>
          <a
            onClick={() =>
              this.props.onAddItem(
                this.state.name,
                this.state.costPerUnit,
                this.state.qty
              )}
            className='btn btn-sm btn-default'
          >
            <span className='glyphicon glyphicon-plus-sign'></span>
          </a>
        </td>
      </tr>
    );
  }
}

export {AddInvoiceItem};
