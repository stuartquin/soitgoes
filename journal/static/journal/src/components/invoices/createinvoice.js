'use strict';
import React from 'react';

import Menu from 'material-ui/Menu';
import MenuItem from 'material-ui/MenuItem';
import RaisedButton from 'material-ui/RaisedButton';
import Popover from 'material-ui/Popover';


class CreateInvoice extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      anchorEl: null,
      open: false
    };
  }

  handleClose() {
    this.setState({
      open: false
    });
  }

  render() {
    const projects = this.props.projects.filter(p => !p.get('archived'));

    return (
      <div className='create-invoice'>

        <RaisedButton
          className='btn-success'
          label='Create New'
          labelPosition='before'
          onTouchTap={(evt) => {
            evt.preventDefault();
            this.setState({
              anchorEl: evt.currentTarget,
              open: true
            });
          }}
        />

        <Popover
          open={this.state.open}
          anchorEl={this.state.anchorEl}
          anchorOrigin={{horizontal: 'left', vertical: 'bottom'}}
          targetOrigin={{horizontal: 'left', vertical: 'top'}}
          onRequestClose={() => this.handleClose()}
        >
          <Menu
            width='256px'
            className='invoice-projects'
            onChange={(evt, project) => {
              this.props.onCreateInvoice(project)
            }}>
          {projects.map((project) => {
            const label = (
              <div className='invoice-project-label'>
                <span>{project.get('name')}</span>
                <span>{project.get('uninvoiced_hours')} hrs</span>
              </div>
            );
            return (
              <MenuItem
                className='invoice-project'
                key={project.get('id')}
                value={project}
                primaryText={label}
              />
            );
          }).toArray()}
          </Menu>
        </Popover>
      </div>
    );
  }
}

export {CreateInvoice};
