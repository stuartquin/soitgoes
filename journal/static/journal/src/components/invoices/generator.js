'use strict';
import React from 'react';
import {Card, CardTitle, CardText} from 'material-ui/Card';
import DatePicker from 'material-ui/DatePicker';
import TextField from 'material-ui/TextField';

import { InvoiceItems } from './invoiceitems';

const timeslipInvoiceItem = (timeslip, project) => {
  const subTotal = project.get('hourly_rate') * timeslip.get('hours');
  return {
    id: timeslip.get('id'),
    details: `${timeslip.get('hours')} hours on ${timeslip.get('date')}`,
    unitPrice: project.get('hourly_rate'),
    subTotal: subTotal
  }
};

const taskInvoiceItem = (task) => {
  return {
    id: task.get('id'),
    details: task.get('name'),
    unitPrice: task.get('cost'),
    subTotal: task.get('cost'),
  }
};


class Generator extends React.Component {
  render() {
    const invoice = this.props.invoice;
    const project = this.props.project;
    const isEditable = this.props.isEditable;

    const timeslipItems = this.props.timeslips.toList().map(t =>
      timeslipInvoiceItem(t, project)
    );
    const taskItems = this.props.tasks.toList().map(taskInvoiceItem);

    let itemViews = [];

    if (timeslipItems.count()) {
      itemViews.push((
        <div key={0}>
          <h4>Tracked Time</h4>
          <InvoiceItems
            isEditable={isEditable}
            items={timeslipItems}
            onDeleteItem={(item) => {
              this.props.onDeleteInvoiceTimeslip(item.id);
            }}
          />
        </div>
      ));
    }

    if (isEditable || taskItems.count()) {
      itemViews.push((
        <div key={1}>
          <h4>Completed Tasks</h4>
          <InvoiceItems
            isEditable={isEditable}
            items={taskItems}
            onDeleteItem={(item) => {
              this.props.onDeleteInvoiceTask(item.id);
            }}
          />
        </div>
      ));
    }

    return (
      <Card className='invoice-generator'>
        <CardText>
          {itemViews}
        </CardText>
      </Card>
    );
  }
}

export {Generator}
