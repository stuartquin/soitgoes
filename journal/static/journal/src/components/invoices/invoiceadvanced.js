import React from 'react';

class InvoiceAdvanced extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      dueDate: props.invoice.get('due_date')
    };
  }

  handleDueDateChange(event) {
    this.setState({
      dueDate: event.target.value
    });
  }

  render() {
    return (
      <div className='invoice-advanced panel panel-default'>
        <div className='invoice-due-date panel-body'>
          <label>
            <span>Due Date:</span>
            <input
              type='date'
              value={this.state.dueDate}
              onChange={(e) => this.handleDueDateChange(e)}
            />
          </label>
          <button className='btn btn-success btn-raised'
            onClick={() => this.props.onUpdate({
              due_date: this.state.dueDate
            })}>
            Update
          </button>
        </div>
      </div>
    );
  }
}

export {InvoiceAdvanced};
