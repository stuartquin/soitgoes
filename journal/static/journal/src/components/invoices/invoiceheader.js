'use strict';
import React from 'react';

import IconButton from 'material-ui/IconButton';
import ActionDelete from 'material-ui/svg-icons/action/delete';
import FileDownload from 'material-ui/svg-icons/file/file-download';
import RaisedButton from 'material-ui/RaisedButton';

import {StateChip} from './state-chip';


const getDefaultName = (invoice, project) => {
  return `${project.get('name')} #${invoice.get('sequence_num')}`;
};

const getStateAction = (invoice, onMarkAsIssued, onMarkAsPaid) => {
  if (!invoice.get('paid_at')) {
    if (invoice.get('issued_at')) {
      return (
        <RaisedButton
          className='btn btn-success'
          label='Set Paid'
          labelPosition='before'
          onTouchTap={onMarkAsPaid}
        />
      );
    } else {
      return (
        <RaisedButton
          className='btn btn-warn'
          label='Issue'
          labelPosition='before'
          onTouchTap={onMarkAsIssued}
        />
      );
    }
  }
};

const getInvoiceActions = (invoice, onDelete) => {
  let actions = [(
    <IconButton
      key={0}
      touch={true}
      tooltipPosition='bottom-center'
      className='btn-default icon-btn-right'
      onTouchTap={onDelete}>
      <ActionDelete />
    </IconButton>
  )];

  if (invoice.get('issued_at')) {
    actions = [(
      <IconButton
        key={1}
        touch={true}
        tooltipPosition='bottom-center'
        className='btn-default icon-btn-right'
        onTouchTap={() => {
          window.open(`/api/invoices/${invoice.get('id')}/pdf`);
        }}>
        <FileDownload />
      </IconButton>
    )].concat(actions);
  }

  return (
    <div className='invoice-header-actions'>
      {actions}
    </div>
  );
};

const InvoiceHeader = (props) => {
  const invoice = props.invoice;
  const project = props.project;
  const contact = props.contact;
  const name = getDefaultName(invoice, project);

  return (
    <div className='invoice-header'>
      <div className='invoice-header-info'>
        <h3>{ name }</h3>
        <span className='text-muted'>
          { contact.get('name') }
        </span>
      </div>
      <div className='invoice-header-options'>
        {getInvoiceActions(invoice, props.onDelete)}
        <div className='invoice-header-actions'>
          <StateChip invoice={invoice} />
          {getStateAction(invoice, props.onMarkAsIssued, props.onMarkAsPaid)}
        </div>
      </div>
    </div>
  );
}


export {InvoiceHeader};
