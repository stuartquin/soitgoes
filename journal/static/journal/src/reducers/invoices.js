'use strict';
import { combineReducers } from 'redux';
import Immutable from 'immutable';

import constants from '../constants';

const getById = (items) => {
  let map = Immutable.OrderedMap();
  items.forEach(item => map = map.set(
    item.id,
    Immutable.fromJS(item)
  ))
  return map;
};

const items = (state = Immutable.OrderedMap(), action) => {
  switch (action.type) {
  case constants.GET_INVOICES_SUCCESS:
    return state.merge(getById(action.invoices));
  default:
    return state;
  }
};

const invoices = combineReducers({
  items
});

export default invoices;
