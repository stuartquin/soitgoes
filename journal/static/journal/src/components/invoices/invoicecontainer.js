'use strict';
import React from 'react';
import {connect} from 'react-redux';
import Immutable from 'immutable';

import {Generator} from './generator';
import {Settings} from './settings';
import {InvoiceHeader} from './invoiceheader';
import {Loading} from '../loading';
import {Confirm} from '../confirm';

import {
  fetchInvoice, deleteInvoice, deleteInvoiceModifier, addInvoiceModifier,
  updateInvoice
} from 'modules/invoice';
import {fetchModifiers} from 'modules/modifier';
import { fetchTasks, updateTask } from 'modules/task';
import { fetchTimeslips, updateTimeslip } from 'modules/timeslip';

class Invoice extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dueDate: null,
      confirmDelete: false
    };
  }

  componentDidMount() {
    this.props.fetchModifiers();
    if (this.props.id !== 'add') {
      this.fetchInvoice().then(() => this.fetchData(this.props.id))
    }
  }

  fetchInvoice() {
    return this.props.fetchInvoice(this.props.id);
  }

  fetchData(id) {
    let promises = [
      this.props.fetchTimeslips(id),
      this.props.fetchTasks(null, null, id)
    ];
    return Promise.all(promises);
  }

  setDueDate(dueDate) {
    this.setState({dueDate});
  }

  handleMarkIssued() {
    const invoice = this.props.invoice;
    this.props.updateInvoice(invoice.get('id'), {
      due_date: this.state.dueDate || invoice.get('due_date'),
      timeslips: this.props.timeslips.map(t => t.get('id')).toJS(),
      status: 'ISSUED'
    });
  }

  handleMarkPaid() {
    const invoice = this.props.invoice;
    this.props.updateInvoice(invoice.get('id'), {
      total_paid: invoice.get('total_due'),
      status: 'PAID'
    });
  }

  render() {
    const invoice = this.props.invoice;
    if (invoice.isEmpty()) {
      return (<Loading />);
    }
    const modifiers = this.props.modifiers;
    const project = this.props.project;
    const contact = this.props.contact;
    const isEditable = !Boolean(invoice.get('issued_at'));
    const dueDate = this.state.dueDate || invoice.get('due_date');

    return (
      <div className='invoice-container'>
        <Confirm
          title='Confirm Delete'
          open={this.state.confirmDelete}
          onConfirm={() => this.props.deleteInvoice(invoice.get('id'))}
          onCancel={() => this.setState({confirmDelete: false})}>
          Are you sure you want to delete?
        </Confirm>

        <div className='header'>
          <InvoiceHeader
            invoice={invoice}
            project={project}
            contact={contact}
            onDelete={() => this.setState({confirmDelete: true})}
            onMarkAsIssued={() => this.handleMarkIssued()}
            onMarkAsPaid={() => this.handleMarkPaid()}
          />
        </div>
        <div className='content'>
          <Settings
            invoice={invoice}
            project={project}
            timeslips={this.props.timeslips}
            tasks={this.props.tasks}
            modifiers={modifiers}
            isEditable={isEditable}
            dueDate={dueDate}
            onAddModifier={(modifier) =>
              this.props.addInvoiceModifier(
                invoice.get('id'),
                modifier.get('id')
              )
            }
            onRemoveModifier={(modifier) =>
              this.props.deleteInvoiceModifier(
                invoice.get('id'),
                modifier.get('id')
              )
            }
            onSetDueDate={(date) => this.setDueDate(date)}
          />
          <Generator
            invoice={invoice}
            project={project}
            timeslips={this.props.timeslips}
            tasks={this.props.tasks}
            modifiers={modifiers}
            isEditable={isEditable}
            onDeleteInvoiceTimeslip={(id) =>
              this.props.updateTimeslip(id, {invoice: null})
            }
            onDeleteInvoiceTask={(id) =>
              this.props.updateTask(id, {invoice: null})
            }
          />
        </div>
      </div>
    );
  }
}

const getInvoiceTimeslips = (invoice, timeslips) => {
  return timeslips.filter((t) => t.get('invoice') === invoice.get('id'));
}

const getInvoiceTasks = (invoice, tasks) => {
  return tasks.filter((t) => t.get('invoice') === invoice.get('id'));
}

const mapStateToProps = (state, { match }) => {
  const params = match.params;
  const id = parseInt(params.id, 10);
  const invoice = state.invoices.items.get(id, Immutable.Map());
  const project = state.projects.items.get(invoice.get('project'), Immutable.Map());
  const contact = state.contacts.items.get(project.get('contact'), Immutable.Map());

  return {
    id,
    invoice,
    project,
    contact,
    modifiers: state.modifiers.items,
    timeslips: getInvoiceTimeslips(invoice, state.timeslips.items),
    tasks: getInvoiceTasks(invoice, state.tasks.items)
  };
};

const actions = {
  fetchInvoice,
  updateInvoice,
  deleteInvoice,

  fetchModifiers,
  deleteInvoiceModifier,
  addInvoiceModifier,

  fetchTasks,
  updateTask,

  fetchTimeslips,
  updateTimeslip
};

const InvoiceContainer = connect(mapStateToProps, actions)(Invoice);
export {InvoiceContainer};
