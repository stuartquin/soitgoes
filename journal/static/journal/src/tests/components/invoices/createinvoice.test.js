// Link.react-test.js
import React from 'react';
import Immutable from 'immutable';
import { CreateInvoice } from '../../../components/invoices/createinvoice';
import renderer from 'react-test-renderer';

it('Defaults to showing create button', () => {
  const component = renderer.create(
    <CreateInvoice
      projects={Immutable.List([])}
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
