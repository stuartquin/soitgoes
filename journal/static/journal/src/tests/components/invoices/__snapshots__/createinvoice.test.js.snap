exports[`test Click button to show project list 1`] = `
<div
  className="create-invoice panel panel-default">
  <div
    className="panel-body">
    <button
      className="btn btn-block btn-success btn-raised"
      onClick={[Function onClick]}>
      New Invoice
    </button>
  </div>
</div>
`;

exports[`test Click button to show project list 2`] = `
<div
  className="create-invoice panel panel-default">
  <div
    className="panel-body">
    <div
      className="new-invoice-project-selector">
      <div
        className="close-action">
        <h5>
          Select Project
        </h5>
        <a
          className="btn btn-default btn-xs"
          onClick={[Function onCancel]}>
          <i
            className="material-icons">
            close
          </i>
          <div
            className="ripple-container" />
        </a>
      </div>
      <hr />
      <ul />
      <hr />
      <div
        className="vat-added">
        <label>
          <input
            checked={true}
            onChange={[Function onChange]}
            type="checkbox" />
          Add VAT
        </label>
      </div>
      <button
        className="btn btn-block btn-success btn-raised"
        onClick={[Function onClick]}>
        Create Invoice
      </button>
    </div>
  </div>
</div>
`;
